graphistry/compute/python_remote.py

```diff
@@ -0,0 +1,103 @@
+from typing import Optional, Union
+from typing_extensions import Literal
+import ast
+import pandas as pd
+import requests
+
+from graphistry.Plottable import Plottable
+
+
+def python_remote(
+    self: Plottable,
+    code: str,
+    api_token: Optional[str] = None,
+    dataset_id: Optional[str] = None,
+    engine: Literal["pandas", "cudf"] = "cudf",
+    validate: bool = True
+) -> Union[Plottable, pd.DataFrame]:
+    """Remotely run Python code on a remote dataset.
+    
+    Uses the latest bound `_dataset_id`, and uploads current dataset if not already bound. Note that rebinding calls of `edges()` and `nodes()` reset the `_dataset_id` binding.
+
+    :param code: Python code that includes a top-level function `def task(g: Plottable) -> Union[str, Dict]`.
+    :type code: str
+
+    :param api_token: Optional JWT token. If not provided, refreshes JWT and uses that.
+    :type api_token: Optional[str]
+
+    :param dataset_id: Optional dataset_id. If not provided, will uplaod current data, store that dataset_id, and run GFQL against that.
```

Comment:
comment references gfql here

Reply:
also could clarify that if a dataset_id exists on the plottable it will be reused, as opposed to reuploading every time python_remote is called with it omitted

Reply:
so this is a bit problematic in that this would trigger re-upload, and not sure if that's 'right':

```python

g2 = g1....

# auto-upload 1
res1 = g2.python_remote(script1)

# auto-upload 2
res2 = g2.python_remote(script2)
```

A few ideas:

* inplace update of `g2` , breaking purity of `v3 = g2.xyz()`
* Some sort of global memoizable trick, so even if `g2._dataset_id` is not mutated on first upload, we can detect `g2` did have a recent upload? Ex: some sort of global weak reference lookup table of last n `g` objects?
* Don't worry about it, and encourage users to do an explicit `g2 = g1.upload()...`

Reply:
didnt consider this... this cant happen in python_remote because that never modifies the plottable, but chain_remote definitely does.

my opinion is that perhaps we should avoid self mutation of plottable and instead return a "clone" (with identical metadata etc) with the updated node and edge lists? this should also better mirror the behaviour of non-remote chain anyway?

Reply:
Does the python endpoint return a Plottable, JSON, or  any? How would I know ahead of time, or how would the python client sniff?

Reply:
for now the python endpoint returns either a string or a json. which one it returns is completely dependent on the return type of the function defined (by the user) in `execute`

Reply:
hmm, it might not be so bad to extend the endpoint to reuse the gfql endpoint's machinery here, i'll take a look

Reply:
(mimetype is cool, that helps!)

Reply:
I agree on avoiding mutation on the base obj, maybe we do some variants like:

```python
# generic helper
_ : Any = g1.remote_python("...", output_mode=...)

# explicit mypy-friendly
g2 = g1.remote_python_g("...")
g2, g1_bound = g1.remote_python_g2("...")
o = g1.remote_python_json("...")
o, g1_bound = g1.remote_python_json2("...")
```